---
title: "Comparison of different methods for calculationg weights in epiregulon with the use of NGS4557 dataset"
author: "Tomasz WÅ‚odarczyk"
date: "`r Sys.Date()`"
output:
  github_document:
    toc: true
    toc_depth: 2
---

```{r setup, include=TRUE}
knitr::opts_chunk$set(
	echo = TRUE,
	message = FALSE,
	warning = FALSE
)
```
# Introduction
This notebook compares across all options for weight computation which are available
in the epiregulon package. As a ground truth NGS4557 dataset is used coming from the study on
the effect of drug treatment on the gene expression of LNCaP and VCaP cell lines. The results are
presented in the form of curves showing the relation between false and true positive rates.

# Highlights

- clusters in methods "corr", "lmfit" and "MI" defined by 'Clusters_Combined' column in the gene expression SCE
- cell line specific weights used for calculation of activity when 'wilcoxon' and 'logFC' methods are used 
- motif scores added
- filtering on motif scores

# Regulon preparation

```{r child = "pruned_regulon_NGS4557.Rmd", warning = FALSE}
```

# Preparation of the data frame to be plotted

```{r}
#library(epiregulon.benchmark)
devtools::load_all("/gstore/project/lineage/tomek/epiregulon.benchmark/")
group_combinations_1 <- expand.grid(c("VCaP", "LNCaP"), c("ARV-110", "Enza"), "DMSO", "down", "AR")
group_combinations_2 <- expand.grid(c("VCaP", "LNCaP"), c("A9690"), "DMSO", "down", c("SMARCA2", "SMARCA4"))
group_combinations <- rbind(group_combinations_1, group_combinations_2)
colnames(group_combinations) <- c("cell_line", "experimental_treatment", "control_treatment", "effect_direction", "transcription_factor")

plot_data <- data.frame()

######
# ArchR project path to be changed
######
```

```{r cache = TRUE}
arg.list <- list()
arg.list$peakMatrix = peakMatrix
arg.list$peak_assay = "counts"
weight_clusters <- c(corr = "Clusters_Combined", MI = "Clusters_Combined", lmfit = "Clusters_Combined", logFC = "Cellline", wilcoxon = "Cellline")
activity_clusters <- c(corr = NA, MI = NA, lmfit = NA, logFC = "Cellline", wilcoxon = "Cellline")
for (method in  c("corr", "MI", "lmfit","logFC","wilcoxon")){
    arg.list$method = method
    partial_data <- prepare_plot_data(regulon = pruned.regulon,
                                      weight.args = arg.list,
                                      group_combinations = group_combinations,
                                      geneExprMatrix.sce = GeneExpressionMatrix,
                                      weight_clusters = weight_clusters[method],
                                      activity_clusters = activity_clusters[method],
                                      path_to_archR_project = "/gstore/project/ar_ligands/AR/scRNAseq/NGS4557_BRG1i/OUTPUT/gp.sa.archr/")
    plot_data <- rbind(plot_data, partial_data)
}

```

```{r}
library(dplyr)
library(ggplot2)

ggplot(plot_data, aes(x=FPR, y=TPR))+
    geom_line(aes(linetype = cell_line, color = tf))+
    facet_grid(treatment~method)+
    xlab("False positive rate")+
    ylab("True positive rate")+
    ggtitle("ROC curves for NGS4557 dataset")


AUC_data <- plot_data %>% group_by(cell_line, treatment, method) %>%
    group_map(function(x,y) cbind(y, data.frame(AUC = calculate_AUC(x$FPR, x$TPR))))
AUC_data <- do.call(rbind, AUC_data)

print(AUC_data)
    
```
