---
title: "Comparison of different methods for calculationg weights in epiregulon with the use of NGS4557 dataset"
author: "Tomasz WÅ‚odarczyk"
date: "`r Sys.Date()`"
output:
  github_document:
    toc: true
    toc_depth: 2
params:
    data_file_paths: ""
    temp_dir: ""
    conda_exe: ""
    virtual_env: ""
    work_dir: ""
    n_cpu: 6
    cell_line: ""
    GRaNIE_tfbs: ""
    motif_score: TRUE
    packages: "c('Epiregulon', 'Pando', 'cellOracle', 'scenicplus', 'GRaNIE', 'FigR')"
    cell_lines: "c('VCaP', 'LNCaP')"
    save_results: TRUE
    save_path: ""
    path_to_ArchR_proj: ""
---

```{r setup, include=TRUE}
knitr::opts_chunk$set(
	echo = TRUE,
	message = FALSE,
	warning = FALSE
)
```

```{r echo = FALSE}
benchmarked_packages <- eval(parse(text = expression(params$packages)))
AR_cell_lines <- eval(parse(text = expression(params$cell_lines)))
save_path <- params$save_path
if(params$save_results && save_path == ""){
    if(!dir.exists(file.path(getwd(), "benchmark_output")))
        dir.create(file.path(getwd(), "benchmark_output"))
    if(!dir.exists(file.path(getwd(), "benchmark_output", "plots")))
        dir.create(file.path(getwd(), "benchmark_output", "plots"))
    save_path <- file.path(getwd(), "benchmark_output")
}
```

# Introduction
This notebook compares across all options for weight computation which are available
in the epiregulon package. As a ground truth NGS4557 dataset is used coming from the study on
the effect of drug treatment on the gene expression of LNCaP and VCaP cell lines. The results are
presented in the form of curves showing the relation between false and true positive rates.

# Highlights

- clusters in methods "corr", "lmfit" and "MI" defined by 'Clusters_Combined' column in the gene expression SCE
- cell line specific weights used for calculation of activity when 'wilcoxon' and 'logFC' methods are used 
- motif scores added
- filtering on motif scores

# Regulon preparation

```{r Epiregulon, child = system.file("reports/pruned_regulon_AR.Rmd", package = "epiregulon.benchmark"),cache=TRUE, cache.lazy==FALSE}
```

# LNCaP
```{r, echo = FALSE}
include_LNCaP <- FALSE
if("LNCaP" %in% AR_cell_lines){
        current_cell_line <- "LNCaP"
        include_LNCaP <- TRUE    
}
```

```{r LNCaP, child = system.file("reports/AR_cell_line.Rmd", package = "epiregulon.benchmark"),cache=TRUE, cache.lazy==FALSE, eval=include_LNCaP}
```

# VCaP
```{r, echo = FALSE}
include_VCaP <- FALSE
if("VCaP" %in% AR_cell_lines){
        current_cell_line <- "VCaP"
        include_VCaP <- TRUE    
}
```

```{r LNCaP, child = system.file("reports/AR_cell_line.Rmd", package = "epiregulon.benchmark"),cache=TRUE, cache.lazy==FALSE, eval=include_VCaP}
```

```{r}
print(sessionInfo())
```

