```{r}
# prepare simulation options
options_ = list(
    rand.seed = 0,
    GRN = GRN,
    num.cells = 5000, # number of cells
    num.cifs = 500,
    tree = phyla, # tree with 9 clusters
    intrinsic.noise = 0.9,
    giv.prob = 0.05,
    augmented.transcr = NULL, # custom parameter
    atac.effect = atac.effect, # effect of regulatory element availability on k_on parameter
    diff.cif.fraction = 0.8, # fraction of CIFs related to the cell development 
    discrete.cif = TRUE, # use discrete mode
    discrete.min.pop.size = 300)
```

```{r eval = TRUE}
library(scMultiSim)
set.seed(10110)
sim_res <- sim_true_counts(options_)

# add noise to gene expression and ATAC-seq data
set.seed(11000)
add_expr_noise(sim_res, protocol = "UMI")
if(params$save_results)
saveRDS(sim_res, file.path(params$save_path, paste0("sim_res_", params$atac.effect, ".rds")))
```


