```{r cache = TRUE, cache.lazy = FALSE}
library(monocle3)
library(cicero)
library(BSgenome.Hsapiens.UCSC.hg38)
library(ArchR)

peakMatrix <- get(paste0("peakMatrix_", current_cell_line))
peak_positions <- rowRanges(peakMatrix)
peakinfo <- data.frame(chr = as.character(seqnames(peak_positions)),
                       bp1 = as.numeric(start(peak_positions)),
                       bp2 = as.numeric(end(peak_positions)))
peakinfo$site_name <- paste(peakinfo$chr, peakinfo$bp1, peakinfo$bp2, sep="_")
row.names(peakinfo) <- peakinfo$site_name
cellinfo = data.frame(cells = colnames(peakMatrix))
rownames(cellinfo) <- cellinfo$cells
indata <- assay(peakMatrix)
rownames(indata) <- rownames(peakinfo)
colnames(indata) <- rownames(cellinfo)

# remove sequences not represented in the reference genome

is.known.sequence <- peakinfo[,1] %in% seqnames(BSgenome.Hsapiens.UCSC.hg38)
peakinfo <- peakinfo[is.known.sequence,]
indata <- indata[is.known.sequence,]


# continue cicero workflow

# create cell_data_set object
indata <- Matrix::Matrix(indata, nrow = nrow(indata),
                         ncol = ncol(indata), dimnames = dimnames(indata))

input_cds <-  suppressWarnings(monocle3::new_cell_data_set(indata,
                                                 cell_metadata = cellinfo,
                                                 gene_metadata = peakinfo))

# sum cells with peak counts above threshold (default = 0)
input_cds <- detect_genes(input_cds)

#Ensure there are no peaks included with zero reads
input_cds <- input_cds[Matrix::rowSums(exprs(input_cds)) != 0,]

set.seed(1010)

input_cds <- detect_genes(input_cds)
input_cds <- estimate_size_factors(input_cds)

# normalization and dimensionality reduction
input_cds <- preprocess_cds(input_cds, method = "LSI")

# UMAP embedding
input_cds <- reduce_dimension(input_cds, reduction_method = 'UMAP',
                              preprocess_method = "LSI")

umap_coords <- reducedDims(input_cds)$UMAP

cicero_cds <- make_cicero_cds(input_cds, reduced_coordinates = umap_coords)

chromosome_lengths <- data.frame(V1 = seqnames(BSgenome.Hsapiens.UCSC.hg38), V2 = seqlengths(BSgenome.Hsapiens.UCSC.hg38))

# Run the main function: calculate co-assessibility between genome sites to predict cis-regulatory interactions
conns <- run_cicero(cicero_cds, chromosome_lengths)

all_peaks <- row.names(exprs(input_cds))

# replace scientific notation with decimal one because it raises the error in python:
# ValueError: invalid literal for int() with base 10: '4.9e+07'
all_peaks <- strsplit(all_peaks, "_")
all_peaks <- do.call(rbind, all_peaks)
all_peaks[,2] <- format(as.numeric(all_peaks[,2]), scientific = FALSE)
all_peaks[,2] <- gsub(" ","",all_peaks[,2])
all_peaks[,3] <- format(as.numeric(all_peaks[,3]), scientific = FALSE)
all_peaks[,3] <- gsub(" ","",all_peaks[,3])
all_peaks <- purrr::pmap(list(all_peaks[,1], all_peaks[,2], all_peaks[,3]), function(x1,x2,x3) paste(x1,x2,x3,sep="_"))
all_peaks <- unlist(all_peaks)
all_peaks <- data.frame(x = all_peaks)
```

```{r base_GRN, cache = TRUE, cache.lazy = FALSE}
library(basilisk)
library(reticulate)
GeneExpressionMatrix = get(paste0("GeneExpressionMatrix_", current_cell_line))

# use lower association_cutoff than value used in the cellOracle tutorial
# otherwise only GATA6 and NKX2-1 will not be recovered
base_GRN <- get_base_GRN(conns, all_peaks)
barcode_tab <- data.frame(barcode = colnames(peakMatrix))
barcode_tab$sample_id <- gsub("(.*)(#.*)", "\\1",barcode_tab$barcode)
barcode_tab$barcode <- gsub(".*#", "",barcode_tab$barcode)
barcode_tab$barcode <- gsub("-", "\\.",barcode_tab$barcode)
barcode_tab$TEST_ARTICLE <- colData(peakMatrix)$TEST_ARTICLE
```
```{r adata, cache = TRUE, cache.lazy = FALSE}
sample_names <- unique(barcode_tab$sample_id)
# adjust sample names order to data paths
new_order <- vapply(sample_names, function(x,y) grep(x,y), params$data_file_paths, FUN.VALUE = numeric(1))
sample_names <- sample_names[new_order]

# Use 5000 instead of 2000 genes since otherwise there an issue caused by the lack of the convergence raised by the sc.tl.diffmap
# function or a waring about many disconnected components in the transformation matrix
adata <- get_annotated_data(barcode_tab,
                            paths_to_data_files = params$data_file_paths,
                            sample_names = sample_names,
                            n_top_genes=2000,
                            variable = "TEST_ARTICLE")
res <- calculate_GRN(adata, base_GRN)
if (params$save_results) saveRDS(res, file.path(save_path, paste0("cellOracle_AR_",current_cell_line, ".rds")))
links <- res[[1]]
louvain_clusters <- res[[2]]
louvain_clusters$barcode <- gsub("(.*)(___)(.*)", paste0("\\3","#","\\1"), louvain_clusters$barcode)
louvain_clusters$barcode <- gsub("\\.","-",louvain_clusters$barcode)
```
