```{r}
library(epiregulon)
library(ArchR)
library(BiocParallel)
library(GenomicRanges)

archR_project_path <- "/gstore/project/ar_ligands/AR/scRNAseq/nonpipeline/OUTPUT/ArchRProject"
proj.all <- loadArchRProject(path = archR_project_path, showLogo = TRUE)
proj.all <- proj.all[grep("SAM24418231|SAM24418230",proj.all$hash_assignment2),]
proj.all <- addCellColData(ArchRProj = proj.all, data  = proj.all$Cell, name = "cell_line", cells = proj.all$cellNames)

###############################
# load gene expression matrix
GeneExpressionMatrix <- getMatrixFromProject(
    ArchRProj = proj.all,
    useMatrix = "GeneExpressionMatrix",
    useSeqnames = NULL,
    verbose = TRUE,
    binarize = FALSE,
    threads = getArchRThreads(),
    logFile = createLogFile("getMatrixFromProject")
)

GeneExpressionMatrix <- ArchRMatrix2SCE(GeneExpressionMatrix, rename="normalizedCounts")
rownames(GeneExpressionMatrix) <- rowData(GeneExpressionMatrix)$name
GeneExpressionMatrix$TEST_ARTICLE <- factor(as.character(GeneExpressionMatrix$TEST_ARTICLE),
                                            levels = c("DMSO", "Enza","ARV110", "A9690"))

# Add embeddings
reducedDim(GeneExpressionMatrix, "UMAP_ATAC") <- getEmbedding(ArchRProj = proj.all,
                                                              embedding = "UMAP_ATAC",
                                                              returnDF = TRUE)[colnames(GeneExpressionMatrix), ]

#load peakMatrix
peakMatrix <- getMatrixFromProject(
    ArchRProj = proj.all,
    useMatrix = "PeakMatrix",
    useSeqnames = NULL,
    verbose = TRUE,
    binarize = FALSE,
    threads = getArchRThreads(),
    logFile = createLogFile("getMatrixFromProject")
)

peakMatrix <- as(peakMatrix, "SingleCellExperiment")
peakMatrix <- peakMatrix[, colnames(GeneExpressionMatrix)]
names(assays(peakMatrix)) <- "counts"

grl <- getTFMotifInfo(genome = "hg38")

selected <- which(proj.all$cell_line == params$cell_line)
proj <- proj.all[proj.all$cell_line == params$cell_line,]
GeneExpressionMatrix <- GeneExpressionMatrix[, selected]
peakMatrix.select <- peakMatrix[, selected]

# peak2gene links

set.seed(1010)

# find cell line specific peaks
cell_peaks <- lapply(unique(proj$hash_assignment2),
                     function(cluster) {readRDS(file.path(archR_project_path, "PeakCalls", paste0(make.names(cluster), "-reproduciblePeaks.gr.rds")))})
cell_peaks <- GenomicRanges::reduce(do.call(c, GRangesList(cell_peaks)))
peaks_overlaps <- findOverlaps(cell_peaks, rowRanges(peakMatrix.select ))
peakMatrix <- peakMatrix.select[unique(subjectHits(peaks_overlaps)),]

# find peak to gene links
p2g <- calculateP2G(peakMatrix = peakMatrix,
                    expMatrix = GeneExpressionMatrix,
                    reducedDim = reducedDim(GeneExpressionMatrix),
                    peak_assay = "counts",
                    exp_assay = "normalizedCounts",
                    cor_cutoff = 0.5
)

# Construct regulons
overlap <- addTFMotifInfo(grl = grl,
                          p2g = p2g,
                          peakMatrix = peakMatrix)


regulon_df_full <- getRegulon(p2g, overlap, aggregate = FALSE)


# prune network
pruned.regulon <- pruneRegulon(regulon = regulon_df_full,
                               expMatrix = GeneExpressionMatrix,
                               exp_assay = "normalizedCounts",
                               peakMatrix = peakMatrix,
                               peak_assay = "counts",
                               prune_value = "pval",
                               clusters = GeneExpressionMatrix$TEST_ARTICLE,
                               exp_cutoff = NULL
)

GeneExpressionMatrix_base <- GeneExpressionMatrix
peakMatrix_base <- peakMatrix
```

