```{r}
library(epiregulon)
library(ArchR)
library(BiocParallel)
library(GenomicRanges)


archR_project_path <- params$path_to_ArchR_proj
proj.all <- loadArchRProject(path = archR_project_path, showLogo = TRUE)
proj.all <- proj.all[!proj.all$hash_assignment2 %in% c(paste0("SAM24425416HTO-",14:16),
                                                       paste0("SAM24425417HTO-",1:10),
                                                       paste0("SAM24428812HTO-",1:6)),]
proj.all$cell_line <- proj.all$Cell

###############################
# load gene expression matrix
GeneExpressionMatrix <- getMatrixFromProject(
    ArchRProj = proj.all,
    useMatrix = "GeneExpressionMatrix",
    useSeqnames = NULL,
    verbose = TRUE,
    binarize = FALSE,
    threads = getArchRThreads(),
    logFile = createLogFile("getMatrixFromProject")
)

GeneExpressionMatrix <- ArchRMatrix2SCE(GeneExpressionMatrix, rename="normalizedCounts")
rownames(GeneExpressionMatrix) <- rowData(GeneExpressionMatrix)$name
GeneExpressionMatrix$TEST_ARTICLE <- factor(as.character(GeneExpressionMatrix$TEST_ARTICLE),
                                            levels = c("DMSO", "Enza","ARV110", "A9690"))

# Add embeddings
reducedDim(GeneExpressionMatrix, "UMAP_ATAC") <- getEmbedding(ArchRProj = proj.all,
                                                              embedding = "UMAP_ATAC",
                                                              returnDF = TRUE)[colnames(GeneExpressionMatrix), ]

#load peakMatrix
peakMatrix <- getMatrixFromProject(
    ArchRProj = proj.all,
    useMatrix = "PeakMatrix",
    useSeqnames = NULL,
    verbose = TRUE,
    binarize = FALSE,
    threads = getArchRThreads(),
    logFile = createLogFile("getMatrixFromProject")
)

peakMatrix <- as(peakMatrix, "SingleCellExperiment")
peakMatrix <- peakMatrix[, colnames(GeneExpressionMatrix)]
names(assays(peakMatrix)) <- "counts"

grl <- getTFMotifInfo(genome = "hg38")

if("LNCaP" %in% AR_cell_lines){
    selected <- which(proj.all$cell_line == "LNCaP")
    proj <- proj.all[proj.all$cell_line == "LNCaP",]
    GeneExpressionMatrix_LNCaP <- GeneExpressionMatrix[, selected]
    peakMatrix.select <- peakMatrix[, selected]
    cell_peaks <- lapply(unique(proj$hash_assignment2),
                         function(cluster) {readRDS(file.path(archR_project_path, "PeakCalls", paste0(make.names(cluster), "-reproduciblePeaks.gr.rds")))})
    cell_peaks <- GenomicRanges::reduce(do.call(c, GRangesList(cell_peaks)))
    peaks_overlaps <- findOverlaps(cell_peaks, rowRanges(peakMatrix.select ))
    peakMatrix_LNCaP <- peakMatrix.select[unique(subjectHits(peaks_overlaps)),]
    
    set.seed(1010, kind ="L'Ecuyer-CMRG")
    
    # find peak to gene links
    p2g <- calculateP2G(peakMatrix = peakMatrix_LNCaP,
                        expMatrix = GeneExpressionMatrix_LNCaP,
                        reducedDim = reducedDim(GeneExpressionMatrix_LNCaP),
                        peak_assay = "counts",
                        exp_assay = "normalizedCounts",
                        cor_cutoff = 0.5
    )
    
    # Construct regulons
    overlap <- addTFMotifInfo(grl = grl,
                              p2g = p2g,
                              peakMatrix = peakMatrix_LNCaP)
    
    
    regulon_df_full_LNCaP <- getRegulon(p2g, overlap, aggregate = FALSE)
    if(params$save_results) saveRDS(regulon_df_full_LNCaP, file.path(save_path, "regulon_full_LNCaP.rds"))
    
    # prune network
    pruned.regulon_LNCaP <- pruneRegulon(regulon = regulon_df_full_LNCaP,
                                   expMatrix = GeneExpressionMatrix_LNCaP,
                                   exp_assay = "normalizedCounts",
                                   peakMatrix = peakMatrix_LNCaP,
                                   peak_assay = "counts",
                                   prune_value = "pval",
                                   clusters = GeneExpressionMatrix_LNCaP$TEST_ARTICLE
    )
    
    if(params$save_results) saveRDS(pruned.regulon_LNCaP, file.path(save_path, "pruned.regulon_LNCaP.rds"))
        
}


if("VCaP" %in% AR_cell_lines){
    selected <- which(proj.all$cell_line == "VCaP")
    proj <- proj.all[proj.all$cell_line == "VCaP",]
    GeneExpressionMatrix_VCaP <- GeneExpressionMatrix[, selected]
    peakMatrix.select <- peakMatrix[, selected]
    cell_peaks <- lapply(unique(proj$hash_assignment2),
                         function(cluster) {readRDS(file.path(archR_project_path, "PeakCalls", paste0(make.names(cluster), "-reproduciblePeaks.gr.rds")))})
    cell_peaks <- GenomicRanges::reduce(do.call(c, GRangesList(cell_peaks)))
    peaks_overlaps <- findOverlaps(cell_peaks, rowRanges(peakMatrix.select ))
    peakMatrix_VCaP <- peakMatrix.select[unique(subjectHits(peaks_overlaps)),]
    
    set.seed(1010, kind ="L'Ecuyer-CMRG")
    
    # find peak to gene links
    p2g <- calculateP2G(peakMatrix = peakMatrix_VCaP,
                        expMatrix = GeneExpressionMatrix_VCaP,
                        reducedDim = reducedDim(GeneExpressionMatrix_VCaP),
                        peak_assay = "counts",
                        exp_assay = "normalizedCounts",
                        cor_cutoff = 0.5
    )
    
    # Construct regulons
    overlap <- addTFMotifInfo(grl = grl,
                              p2g = p2g,
                              peakMatrix = peakMatrix_VCaP)
    
    
    regulon_df_full_VCaP <- getRegulon(p2g, overlap, aggregate = FALSE)
    if(params$save_results) saveRDS(regulon_df_full_VCaP, file.path(save_path, "regulon_full_VCaP.rds"))
    
    # prune network
    pruned.regulon_VCaP <- pruneRegulon(regulon = regulon_df_full_VCaP,
                                   expMatrix = GeneExpressionMatrix_VCaP,
                                   exp_assay = "normalizedCounts",
                                   peakMatrix = peakMatrix_VCaP,
                                   peak_assay = "counts",
                                   prune_value = "pval",
                                   clusters = GeneExpressionMatrix_VCaP$TEST_ARTICLE
    )
    
    if(params$save_results) saveRDS(pruned.regulon_VCaP, file.path(save_path, "pruned.regulon_VCaP.rds"))
        
}

```

