```{r}
library(scran)
library(scater)
library(epiregulon)
library(ArchR)
library(BiocParallel)

#checkdata
archR_project_path <- "/gstore/project/ar_ligands/AR/scRNAseq/NGS4557_BRG1i/OUTPUT/gp.sa.archr/"
proj <- loadArchRProject(path = archR_project_path, showLogo = TRUE)

# load gene expression matrix
GeneExpressionMatrix <- getMatrixFromProject(
    ArchRProj = proj,
    useMatrix = "GeneExpressionMatrix",
    useSeqnames = NULL,
    verbose = TRUE,
    binarize = FALSE,
    threads = getArchRThreads(),
    logFile = createLogFile("getMatrixFromProject")
)

GeneExpressionMatrix$cell_line <- sapply(strsplit(GeneExpressionMatrix$hash_assignment, split = "_"), "[",2)

GeneExpressionMatrix <- ArchRMatrix2SCE(GeneExpressionMatrix, rename="normalizedCounts")
rownames(GeneExpressionMatrix) <- rowData(GeneExpressionMatrix)$name
GeneExpressionMatrix$hash_assignment <- unlist(lapply(GeneExpressionMatrix$hash_assignment, 
                                                        function(x) gsub("-", ".", x)))

TEST_ARTICLE <- setNames(c("Enza", "DMSO", "DMSO", "ARV110", "Enza", "ARV110", "A9690", "A9690"), unique(GeneExpressionMatrix$hash_assignment))

GeneExpressionMatrix$TEST_ARTICLE <- TEST_ARTICLE[GeneExpressionMatrix$hash_assignment]

reducedDim(GeneExpressionMatrix, "UMAP_ATAC") <- getEmbedding(ArchRProj = proj,
                                                              embedding = "UMAP_ATAC",
                                                              returnDF = TRUE)[colnames(GeneExpressionMatrix), ]

proj <- addCellColData(proj, data = GeneExpressionMatrix$cell_line, name = "cell_line",
                       cells = proj$cellNames)

#load peakMatrix
peakMatrix <- getMatrixFromProject(
    ArchRProj = proj,
    useMatrix = "PeakMatrix",
    useSeqnames = NULL,
    verbose = TRUE,
    binarize = FALSE,
    threads = getArchRThreads(),
    logFile = createLogFile("getMatrixFromProject")
)

peakMatrix <- as(peakMatrix, "SingleCellExperiment")
peakMatrix <- peakMatrix[, colnames(GeneExpressionMatrix)]
names(assays(peakMatrix)) <- "counts"

########################
# epiregulon
# get TF binding
grl <- getTFMotifInfo(genome = "hg38")


selected <- which(proj$cell_line == params$cell_line)
proj <- proj[proj$cell_line == params$cell_line,]
GeneExpressionMatrix <- GeneExpressionMatrix[, selected]
peakMatrix <- peakMatrix[, selected]


# peak2gene links

set.seed(1010)

# find cell line specific peaks
cell_peaks <- lapply(unique(proj$hash_assignment),
                     function(cluster) {readRDS(file.path(archR_project_path, "PeakCalls", paste0(make.names(cluster), "-reproduciblePeaks.gr.rds")))})
cell_peaks <- GenomicRanges::reduce(do.call(c, GRangesList(cell_peaks)))
peaks_overlaps <- findOverlaps(cell_peaks, rowRanges(peakMatrix))
peakMatrix <- peakMatrix[subjectHits(peaks_overlaps),]

# find peak to gene links
p2g <- calculateP2G(peakMatrix = peakMatrix,
                    expMatrix = GeneExpressionMatrix,
                    reducedDim = reducedDim(GeneExpressionMatrix),
                    peak_assay = "counts",
                    exp_assay = "normalizedCounts",
                    cor_cutoff = 0.5
)


# Construct regulons
overlap <- addTFMotifInfo(grl = grl,
                          p2g = p2g,
                          peakMatrix = peakMatrix)



regulon_df_full <- getRegulon(p2g, overlap, aggregate = FALSE)




# prune network
pruned.regulon <- pruneRegulon(regulon = regulon_df_full,
                               expMatrix = GeneExpressionMatrix,
                               exp_assay = "normalizedCounts",
                               peakMatrix = peakMatrix,
                               peak_assay = "counts",
                               prune_value = "pval",
                               clusters = GeneExpressionMatrix$TEST_ARTICLE,
                               exp_cutoff = NULL
)



```

