---
title: "Comparison of different methods for calculationg weights in epiregulon with the use of reprogram-seq dataset"
date: "`r Sys.Date()`"
output:
  github_document:
    toc: true
    toc_depth: 2
params:
    treatments: list(GATA6 = NULL, NKX2.1 = NULL, FOXA2 = NULL)
    cluster_column: "Clusters"
    data_file_paths: ""
    temp_dir: ""
    work_dir: ""
    n_cpu: 12
    group_variable: "HTO"
    GRaNIE_tfbs: ""
    save_path: ""
    packages: "c('Epiregulon', 'Pando', 'scenicplus', 'GRaNIE', 'FigR')"
    save_results: TRUE
    motif_db_dir: ""
---

```{r setup, include=TRUE, cache = TRUE, cache.lazy=FALSE}
knitr::opts_chunk$set(
	echo = TRUE,
	message = FALSE,
	warning = FALSE
)
```
# Introduction
This report presents comparison across all method options for weight computation which are available
in the `epiregulon` package. As a ground truth the reprogram-seq dataset is used. It was generated as a result of the studies in which cells were modified to overexpress a selected transcription factor. The results are
presented in the form of curves showing the relation between false and true positive rates.

# Regulon preparation
```{r child = system.file("reports/get_matrices_reprogram-seq.Rmd", package = "epiregulon.benchmark"), cache = TRUE, cache.lazy = FALSE}
```

```{r child = system.file("reports/pruned_regulon_reprogram-seq.Rmd", package = "epiregulon.benchmark"), cache=TRUE, cache.lazy=FALSE}
```

```{r echo=FALSE}
treatments <- eval(parse(text = enquote(params$treatments)))
benchmarked_packages <- eval(parse(text = enquote(params$packages)))
save_path <- params$save_path
if(params$save_results && save_path == ""){
    if(!dir.exists(file.path(getwd(), "benchmark_output")))
        dir.create(file.path(getwd(), "benchmark_output"))
    if(!dir.exists(file.path(getwd(), "benchmark_output", "plots")))
        dir.create(file.path(getwd(), "benchmark_output", "plots"))
    save_path <- file.path(getwd(), "benchmark_output")
}
if(!dir.exists(file.path(save_path, "plots")))
    dir.create(file.path(save_path, "plots"))
```

# Comparison of the performance of different weight methods
```{r cache = TRUE, cache.lazy=FALSE}
library(epiregulon.benchmark)
group_combinations <- expand.grid(NA, names(treatments), NA, "up", "")
colnames(group_combinations) <- c("cell_line", "experimental_treatment", "control_treatment", "effect_direction", "transcription_factor")
control_treatment <- c()
for(i in seq_along(treatments)){
    if(is.null(treatments[[i]]))
        control_treatment[i] <- get_complementary_names(names(treatments)[i], unique(GeneExpressionMatrix$hash_assignment))
    else
        control_treatment[i] <- paste0(params$treatments[[i]], collapse = "|")
}
group_combinations$control_treatment <- control_treatment
group_combinations$transcription_factor <- unlist(lapply(group_combinations$experimental_treatment, function(x) map_treatment_to_tf(x, regulon)))
```


```{r results = "hide", cache = TRUE, cache.lazy=FALSE, message=FALSE, warning=FALSE}
library(ggplot2)
regulon <- regulon[regulon$tf %in% c("NKX2-1", "GATA6", "FOXA2"),]
plot_data <- data.frame()
arg.list <- list()
arg.list$peakMatrix = PeakMatrix
arg.list$peak_assay = "counts"
arg.list$exp_assay = "normalizedCounts"
arg.list$tf_re.merge = TRUE
weight_clusters  <- as.list(setNames(rep("Clusters", 2), c("corr", "MI")))
for (method in  c("wilcoxon", "corr", "MI")){
    arg.list$method = method
    partial_data <- prepare_plot_data(regulon = regulon,
                                      weight.args = arg.list,
                                      group_combinations = group_combinations,
                                      geneExprMatrix.sce = GeneExpressionMatrix,
                                      weight_clusters = weight_clusters[[method]],
                                      activity_clusters = NA,
                                      motif_score = FALSE)
    plot_data <- rbind(plot_data, partial_data)
}

plot_data$method[plot_data$method %in% c("corr", "MI")] <- paste0(plot_data$method[plot_data$method %in% c("corr", "MI")], " (tf, re)")

arg.list$tf_re.merge = FALSE
for (method in  c("corr", "MI")){
    arg.list$method = method
    partial_data <- prepare_plot_data(regulon = regulon,
                                      weight.args = arg.list,
                                      group_combinations = group_combinations,
                                      geneExprMatrix.sce = GeneExpressionMatrix,
                                      weight_clusters = weight_clusters[[method]],
                                      activity_clusters = NA,
                                      motif_score = FALSE)
    plot_data <- rbind(plot_data, partial_data)
}
plot_data$method[plot_data$method %in% c("corr", "MI")] <- paste0(plot_data$method[plot_data$method %in% c("corr", "MI")], " (tf)")
plot_data$method <- factor(plot_data$method, levels = c("wilcoxon", "corr (tf, re)", "MI (tf, re)", "corr (tf)", "MI (tf)"))
plot_data$treatment <- factor(plot_data$treatment, levels = c("NKX2.1", "HTO8_NKX2.1", "GATA6", "FOXA2"))
plot_data <- plot_data[plot_data$treatment!="NKX2.1",]
plot_data$treatment[plot_data$treatment=="HTO8_NKX2.1"] = "NKX2.1"
#plot_data$treatment <- droplevels(plot_data$treatment)
AUC_data <- plot_data %>% group_by(cell_line, treatment, method, tf) %>%
    group_map(function(x,y) cbind(y, data.frame(AUC = calculate_AUC(x$FPR, x$TPR))))
AUC_data <- do.call(rbind, AUC_data)
ggplot(plot_data, aes(x=FPR, y=TPR))+
    geom_line()+
    facet_grid(treatment~method)+
    xlab("False positive rate")+
    ylab("True positive rate")+
    ggtitle("ROC curves for reprogram-seq dataset")+
    geom_text(data = AUC_data, aes(x=0.8, y=0.3, label = sprintf("%.3f", AUC)))+
    theme(panel.border = element_rect(color = "black", fill = NA, linewidth = 1),
          panel.background = element_rect(fill = "white", colour="white"),
          plot.title = element_text(hjust = 0.5))

print(AUC_data)

if(params$save_results){
    ggsave(file.path(save_path, "plots", "Reprogram-seq_weight_methods.pdf"), width = 11, height = 6.6)
    write.csv(AUC_data, file.path(save_path, "AUC_reprogram-seq_by_weight_method.csv"))
    write.csv(plot_data, file.path(save_path, "plot_data_reprogram-seq_by_weight_method.csv"))
}
```

```{r echo=FALSE}
GRN_list <- list()
```

```{r echo = FALSE}
eval_package = FALSE
if("Epiregulon" %in% benchmarked_packages) eval_package = TRUE
```

```{r cache = TRUE, cache.lazy=FALSE, eval=eval_package}
weight.args <- list(clusters = as.vector(GeneExpressionMatrix[[params$cluster_column]]), regulon = regulon,
                    expMatrix = GeneExpressionMatrix, peakMatrix = PeakMatrix, peak_assay = "counts",
                    exp_assay = "normalizedCounts", method = "wilcoxon")
# epiregulon workflow
regulon.w <- do.call(addWeights, weight.args)
GRN_list[["Epiregulon"]] <- regulon.w
if(params$save_results)
    saveRDS(regulon.w, file.path(save_path, "Epiregulon_reprogram.rds"))
```

```{r echo = FALSE}
eval_package = FALSE
if("Pando" %in% benchmarked_packages) eval_package = TRUE
```

```{r child = system.file("reports/Pando_reprogram-seq.Rmd", package = "epiregulon.benchmark"), cache = FALSE, cache.lazy=FALSE, eval=eval_package}
```

```{r cache = TRUE, cache.lazy=FALSE}
Seurat_obj  <- Seurat::RenameCells(Seurat_obj, new.names = paste0("reprogram#", colnames(Seurat_obj))) 
GRN_list[["Pando"]] <- Seurat_obj 
```


```{r echo = FALSE}
eval_package = FALSE
if("FigR" %in% benchmarked_packages) eval_package = TRUE
```

```{r child = system.file("reports/FigR_reprogram-seq.Rmd", package = "epiregulon.benchmark"), cache = TRUE, cache.lazy=FALSE,message=FALSE, eval=eval_package}
```

```{r cache = TRUE, cache.lazy=FALSE}
GRN_list[["FigR"]] <- FigR_GRN
```

```{r echo = FALSE}
eval_package = FALSE
if("GRaNIE" %in% benchmarked_packages) eval_package = TRUE
```

```{r child = system.file("reports/GRaNIE_reprogram-seq.Rmd", package = "epiregulon.benchmark"), cache = TRUE, cache.lazy=FALSE,message=FALSE, eval=eval_package}
```

```{r cache = TRUE, cache.lazy=FALSE, eval=eval_package}
GRN_list[["GRaNIE"]] <- GRN_connections.all
```

```{r echo = FALSE}
eval_package = FALSE
if("scenicplus" %in% benchmarked_packages) eval_package = TRUE
```

```{r child = system.file("reports/Scenic_plus_reprogram-seq.Rmd", package = "epiregulon.benchmark"), cache = TRUE, cache.lazy=FALSE, eval=eval_package}
```

```{r eval=eval_package}
GRN_list[["scenic_plus"]] <- scenic_plus_res
```

```{r cache = TRUE, cache.lazy=FALSE, message=FALSE}
matrices_list <- list()
for(i in seq_along(GRN_list)){
    GRN <- GRN_list[[i]]
    method <- names(GRN_list)[i]
    activity.matrix <- get_activity_matrix(GRN = GRN,
                                           method = method,
                                           geneExprMatrix.sce = GeneExpressionMatrix,
                                           tfs = c("GATA6", "NKX2-1", "FOXA2"),
                                           exp_assay = "normalizedCoutns")
  matrices_list[[i]] <- activity.matrix 
}

names(matrices_list) <- names(GRN_list)
matrices_list <- c(matrices_list, GRN_list[["scenic_plus"]])
```

```{r cache = FALSE, cache.lazy=FALSE, message=FALSE, fig.dim=c(6,4)}
library(ggplot2)
res <- plotDataFromActivity(matrices_list, tf=group_combinations[1,"transcription_factor"],
                            labels = GeneExpressionMatrix$hash_assignment,
                            positive_elements_label=group_combinations[1,"experimental_treatment"],
                            negative_elements_label=group_combinations[1,"control_treatment"],
                            GeneExpressionMatrix,
                            experimental_treatment = group_combinations[1,"experimental_treatment"],
                            title = "Reprogram-seq, induced GATA6 overexpression\nGATA6 activity")
plot(res[[1]])
if(params$save_results)
    ggsave(file.path(save_path, "plots", "Reprogram-seq_by_package_GATA6.pdf"), width = 6, height = 4)


res <- plotDataFromActivity(matrices_list, tf=group_combinations[2,"transcription_factor"],
                            labels = GeneExpressionMatrix$hash_assignment,
                            positive_elements_label=group_combinations[2,"experimental_treatment"],
                            negative_elements_label=group_combinations[2,"control_treatment"],
                            GeneExpressionMatrix,
                            experimental_treatment = group_combinations[2,"experimental_treatment"],
                            title = "Reprogram-seq, induced FOXA2 overexpression\nFOXA2 activity")
plot(res[[1]])

if(params$save_results)
    ggsave(file.path(save_path, "plots", "Reprogram-seq_by_package_FOXA2.pdf"), width = 6, height = 4)

res <- plotDataFromActivity(matrices_list, tf=group_combinations[3,"transcription_factor"],
                            labels = GeneExpressionMatrix$hash_assignment,
                            positive_elements_label=group_combinations[3,"experimental_treatment"],
                            negative_elements_label=group_combinations[3,"control_treatment"],
                            GeneExpressionMatrix,
                            experimental_treatment = group_combinations[3,"experimental_treatment"],
                            title = "Reprogram-seq, induced NKX2-1 overexpression\nNKX2-1 activity")

plot(res[[1]])

if(params$save_results)
    ggsave(file.path(save_path, "plots", "Reprogram-seq_by_package_NKX2-1.pdf"), width = 6, height = 4)

```

# Number of recovered transcription factors
```{r fig.dim=c(4.5,4)}
res <- c(Epiregulon = 0, Pando = 0, FigR = 0, GRaNIE = 0)
for(i in seq_along(GRN_list)){
    current_package <- names(GRN_list)[i]
    if(current_package == "Epiregulon")
        res["Epiregulon"] <- length(unique(GRN_list$Epiregulon$tf))
    else if(current_package == "Pando"){
        res["Pando"] <- length(unique(GRN_list$Pando@grn@networks$glm_network@coefs$tf))
    }
    else if(current_package == "FigR")
        res["FigR"] <- length(unique(GRN_list$FigR$Motif))
    else if(current_package == "GRaNIE")
        res["GRaNIE"] <- length(unique(GRN_list$GRaNIE$TF.name))
}

for(score.matrix.name in c("scenic_plus_pos_gene", "scenic_plus_neg_gene",
                           "scenic_plus_pos_region", "scenic_plus_neg_region")){
    res[score.matrix.name] <- nrow(matrices_list[[score.matrix.name]])
}
res <- data.frame(n_tf = res, package = names(res))

res <- res[res$n_tf>0,]

colors <- c(Epiregulon = "red", FigR = "black", Pando = "blue", GRaNIE = "green1",
             scenic_plus_neg_gene = "grey", scenic_plus_pos_gene = "orange",
            scenic_plus_pos_region = "green4", scenic_plus_neg_region = "tan4",
            "Gene expression" = "darkslategray2")

colors <- colors[unique(res$package)]

library(ggplot2)
ggplot(res, aes(y = n_tf, x = package, fill = package))+
    geom_bar(stat = "identity", width = 0.6)+
    scale_color_manual(values = colors)+
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size =11, lineheight = 0.8, family = "sans"),
          panel.grid.major.y = element_line(colour = "white", linewidth =0.25),
          panel.grid.minor.y = element_line(colour = "white", linewidth =0.1),
          panel.grid.major.x = element_blank(),
          panel.background = element_rect(fill = "white", colour="black"),
          plot.title = element_text(hjust = 0.5, size = 13, family = "sans"),
          legend.position = "none")+
    ylab("Number of transcription factors found")+
    xlab("Package")+
    ggtitle("The number of transcription factors found\nwith the use of different GRN inference tools")+
    scale_x_discrete(labels=c("Epiregulon" = "Epiregulon", "Pando" = "Pando", 
                              "FigR" = "FigR", "GRaNIE" = "GRaNIE",
                              "scenic_plus_pos_gene" = "scenicplus\ngene-based\npositive",
                              "scenic_plus_neg_gene" = "scenicplus\ngene-baned\nnegative",
                              "scenic_plus_pos_region" = "scenicplus\nregion-based\npositive",
                              "scenic_plus_neg_region" = "scenicplus\nregion-baned\nnegative"))
if(params$save_results)
    ggsave(file.path(save_path, "plots", "TF_numbers_reprogram-seq.pdf"), width = 4.5, height = 4)
```

```{r}
sessionInfo()
```
